### ISSUE TITLE
Build failure with GHC 8

### ISSUE THREAD

**Issue by snoyberg** [well-typed/hackage-security#164]
```
> /tmp/stackage-build8/hackage-security-0.5.1.0$ runghc -clear-package-db -global-package-db -package-db=/home/stackage/work/builds/nightly/pkgdb Setup configure --enable-tests --package-db=clear --package-db=global --package-db=/home/stackage/work/builds/nightly/pkgdb --libdir=/home/stackage/work/builds/nightly/lib --bindir=/home/stackage/work/builds/nightly/bin --datadir=/home/stackage/work/builds/nightly/share --libexecdir=/home/stackage/work/builds/nightly/libexec --sysconfdir=/home/stackage/work/builds/nightly/etc --docdir=/home/stackage/work/builds/nightly/doc/hackage-security-0.5.1.0 --htmldir=/home/stackage/work/builds/nightly/doc/hackage-security-0.5.1.0 --haddockdir=/home/stackage/work/builds/nightly/doc/hackage-security-0.5.1.0 --flags=
Configuring hackage-security-0.5.1.0...
> /tmp/stackage-build8/hackage-security-0.5.1.0$ runghc -clear-package-db -global-package-db -package-db=/home/stackage/work/builds/nightly/pkgdb Setup build
Building hackage-security-0.5.1.0...
Preprocessing library hackage-security-0.5.1.0...
[ 1 of 42] Compiling Prelude          ( src/Prelude.hs, dist/build/Prelude.o )
[ 2 of 42] Compiling Hackage.Security.Util.Checked ( src/Hackage/Security/Util/Checked.hs, dist/build/Hackage/Security/Util/Checked.o )

src/Hackage/Security/Util/Checked.hs:51:1: warning: [-Wredundant-constraints]
    • Redundant constraint: Throws e
    • In the type signature for:
           throwChecked :: (Exception e, Throws e) => e -> IO a
[ 3 of 42] Compiling Hackage.Security.Util.Lens ( src/Hackage/Security/Util/Lens.hs, dist/build/Hackage/Security/Util/Lens.o )
[ 4 of 42] Compiling Hackage.Security.Util.Pretty ( src/Hackage/Security/Util/Pretty.hs, dist/build/Hackage/Security/Util/Pretty.o )
[ 5 of 42] Compiling Hackage.Security.Util.Path ( src/Hackage/Security/Util/Path.hs, dist/build/Hackage/Security/Util/Path.o )
[ 6 of 42] Compiling Hackage.Security.Util.IO ( src/Hackage/Security/Util/IO.hs, dist/build/Hackage/Security/Util/IO.o )
[ 7 of 42] Compiling Hackage.Security.Client.Verify ( src/Hackage/Security/Client/Verify.hs, dist/build/Hackage/Security/Client/Verify.o )
[ 8 of 42] Compiling Text.JSON.Canonical ( src/Text/JSON/Canonical.hs, dist/build/Text/JSON/Canonical.o )
[ 9 of 42] Compiling Hackage.Security.TUF.Paths ( src/Hackage/Security/TUF/Paths.hs, dist/build/Hackage/Security/TUF/Paths.o )

src/Hackage/Security/TUF/Paths.hs:40:1: warning: [-Wredundant-constraints]
    • Redundant constraint: FsRoot root
    • In the type signature for:
           anchorRepoPathLocally :: FsRoot root =>
                                    Path root -> RepoPath -> Path root

src/Hackage/Security/TUF/Paths.hs:71:1: warning: [-Wredundant-constraints]
    • Redundant constraint: FsRoot root
    • In the type signature for:
           anchorCachePath :: FsRoot root =>
                              Path root -> CachePath -> Path root
[10 of 42] Compiling Hackage.Security.TUF.Layout.Cache ( src/Hackage/Security/TUF/Layout/Cache.hs, dist/build/Hackage/Security/TUF/Layout/Cache.o )
[11 of 42] Compiling Hackage.Security.TUF.Layout.Repo ( src/Hackage/Security/TUF/Layout/Repo.hs, dist/build/Hackage/Security/TUF/Layout/Repo.o )
[12 of 42] Compiling Hackage.Security.Util.JSON ( src/Hackage/Security/Util/JSON.hs, dist/build/Hackage/Security/Util/JSON.o )
[13 of 42] Compiling Hackage.Security.Util.Base64 ( src/Hackage/Security/Util/Base64.hs, dist/build/Hackage/Security/Util/Base64.o )
[14 of 42] Compiling Hackage.Security.Util.Stack ( src/Hackage/Security/Util/Stack.hs, dist/build/Hackage/Security/Util/Stack.o )
[15 of 42] Compiling Hackage.Security.Util.TypedEmbedded ( src/Hackage/Security/Util/TypedEmbedded.hs, dist/build/Hackage/Security/Util/TypedEmbedded.o )
[16 of 42] Compiling Hackage.Security.Client.Formats ( src/Hackage/Security/Client/Formats.hs, dist/build/Hackage/Security/Client/Formats.o )

src/Hackage/Security/Client/Formats.hs:116:1: warning: [-Woverlapping-patterns]
    Pattern match is redundant
    In an equation for ‘formatsLookup’: formatsLookup _ _ = ...
[17 of 42] Compiling Hackage.Security.Util.Some ( src/Hackage/Security/Util/Some.hs, dist/build/Hackage/Security/Util/Some.o )
[18 of 42] Compiling Hackage.Security.Key ( src/Hackage/Security/Key.hs, dist/build/Hackage/Security/Key.o )
[19 of 42] Compiling Hackage.Security.Key.Env ( src/Hackage/Security/Key/Env.hs, dist/build/Hackage/Security/Key/Env.o )
[20 of 42] Compiling Hackage.Security.JSON ( src/Hackage/Security/JSON.hs, dist/build/Hackage/Security/JSON.o )
[21 of 42] Compiling Hackage.Security.TUF.Common ( src/Hackage/Security/TUF/Common.hs, dist/build/Hackage/Security/TUF/Common.o )
[22 of 42] Compiling Hackage.Security.TUF.FileInfo ( src/Hackage/Security/TUF/FileInfo.hs, dist/build/Hackage/Security/TUF/FileInfo.o )
[23 of 42] Compiling Hackage.Security.TUF.FileMap ( src/Hackage/Security/TUF/FileMap.hs, dist/build/Hackage/Security/TUF/FileMap.o )
[24 of 42] Compiling Hackage.Security.TUF.Header ( src/Hackage/Security/TUF/Header.hs, dist/build/Hackage/Security/TUF/Header.o )
[25 of 42] Compiling Hackage.Security.TUF.Signed ( src/Hackage/Security/TUF/Signed.hs, dist/build/Hackage/Security/TUF/Signed.o )
[26 of 42] Compiling Hackage.Security.TUF.Mirrors ( src/Hackage/Security/TUF/Mirrors.hs, dist/build/Hackage/Security/TUF/Mirrors.o )
[27 of 42] Compiling Hackage.Security.TUF.Snapshot ( src/Hackage/Security/TUF/Snapshot.hs, dist/build/Hackage/Security/TUF/Snapshot.o )
[28 of 42] Compiling Hackage.Security.TUF.Timestamp ( src/Hackage/Security/TUF/Timestamp.hs, dist/build/Hackage/Security/TUF/Timestamp.o )
[29 of 42] Compiling Hackage.Security.TUF.Patterns ( src/Hackage/Security/TUF/Patterns.hs, dist/build/Hackage/Security/TUF/Patterns.o )
[30 of 42] Compiling Hackage.Security.TUF.Targets ( src/Hackage/Security/TUF/Targets.hs, dist/build/Hackage/Security/TUF/Targets.o )
[31 of 42] Compiling Hackage.Security.TUF.Layout.Index ( src/Hackage/Security/TUF/Layout/Index.hs, dist/build/Hackage/Security/TUF/Layout/Index.o )
[32 of 42] Compiling Hackage.Security.TUF.Root ( src/Hackage/Security/TUF/Root.hs, dist/build/Hackage/Security/TUF/Root.o )
[33 of 42] Compiling Hackage.Security.TUF ( src/Hackage/Security/TUF.hs, dist/build/Hackage/Security/TUF.o )
[34 of 42] Compiling Hackage.Security.Server ( src/Hackage/Security/Server.hs, dist/build/Hackage/Security/Server.o )
[35 of 42] Compiling Hackage.Security.Trusted.TCB ( src/Hackage/Security/Trusted/TCB.hs, dist/build/Hackage/Security/Trusted/TCB.o )
[36 of 42] Compiling Hackage.Security.Trusted ( src/Hackage/Security/Trusted.hs, dist/build/Hackage/Security/Trusted.o )
[37 of 42] Compiling Hackage.Security.Client.Repository ( src/Hackage/Security/Client/Repository.hs, dist/build/Hackage/Security/Client/Repository.o )
[38 of 42] Compiling Hackage.Security.Client ( src/Hackage/Security/Client.hs, dist/build/Hackage/Security/Client.o )

src/Hackage/Security/Client.hs:113:5: warning: [-Wredundant-constraints]
    • Redundant constraints: (Throws VerificationError,
                              Throws SomeRemoteError)
    • In the type signature for:
           limitIterations :: (Throws VerificationError,
                               Throws SomeRemoteError) =>
                              VerificationHistory -> IO HasUpdates
      In an equation for ‘checkForUpdates’:
          checkForUpdates rep@(Repository {..}) mNow
            = withMirror rep $ limitIterations []
            where
                maxNumIterations :: Int
                maxNumIterations = 5
                limitIterations ::
                  (Throws VerificationError, Throws SomeRemoteError) =>
                  VerificationHistory -> IO HasUpdates
                limitIterations history
                  | length history >= maxNumIterations
                  = throwChecked $ VerificationErrorLoop (reverse history)
                limitIterations history
                  = do { cachedInfo <- getCachedInfo rep;
                         .... }
                  where
                      attemptNr :: AttemptNr
                      attemptNr = fromIntegral $ length history
                ....

src/Hackage/Security/Client.hs:336:1: warning: [-Wredundant-constraints]
    • Redundant constraint: Applicative m
    • In the type signature for:
           getCachedInfo :: (Applicative m, MonadIO m) =>
                            Repository down -> m CachedInfo

src/Hackage/Security/Client.hs:357:1: warning: [-Wredundant-constraints]
    • Redundant constraint: Applicative m
    • In the type signature for:
           readLocalFile :: (FromJSON ReadJSON_Keys_Layout (Signed a),
                             MonadIO m, Applicative m) =>
                            Repository down -> KeyEnv -> CachedFile -> m (Maybe (Trusted a))
[39 of 42] Compiling Hackage.Security.Client.Repository.Cache ( src/Hackage/Security/Client/Repository/Cache.hs, dist/build/Hackage/Security/Client/Repository/Cache.o )

src/Hackage/Security/Client/Repository/Cache.hs:65:5: warning: [-Wredundant-constraints]
    • Redundant constraint: typ ~ Binary
    • In the type signature for:
           unzipIndex :: typ ~ Binary => IO ()
      In an equation for ‘cacheRemoteFile’:
          cacheRemoteFile cache downloaded f isCached
            = do { go f isCached;
                   case isCached of {
                     CacheIndex -> rebuildTarIndex cache
                     _otherwise -> return () } }
            where
                go :: Format f -> IsCached typ -> IO ()
                go _ DontCache = return ()
                go FUn (CacheAs file) = copyTo (cachedFilePath cache file)
                go FGz CacheIndex
                  = copyTo (cachedIndexPath cache FGz) >> unzipIndex
                go _ _ = error "cacheRemoteFile: unexpected case"
                copyTo :: Path Absolute -> IO ()
                copyTo fp
                  = do { createDirectoryIfMissing True (takeDirectory fp);
                         .... }
                ....
[40 of 42] Compiling Hackage.Security.Client.Repository.HttpLib ( src/Hackage/Security/Client/Repository/HttpLib.hs, dist/build/Hackage/Security/Client/Repository/HttpLib.o )
[41 of 42] Compiling Hackage.Security.Client.Repository.Local ( src/Hackage/Security/Client/Repository/Local.hs, dist/build/Hackage/Security/Client/Repository/Local.o )
[42 of 42] Compiling Hackage.Security.Client.Repository.Remote ( src/Hackage/Security/Client/Repository/Remote.hs, dist/build/Hackage/Security/Client/Repository/Remote.o )

src/Hackage/Security/Client/Repository/Remote.hs:365:5: warning: [-Wredundant-constraints]
    • Redundant constraint: Throws SomeRemoteError
    • In the type signature for:
           go :: Throws SomeRemoteError =>
                 DownloadMethod fs typ
                 -> Verify (Some (HasFormat fs), RemoteTemp typ)
      In an equation for ‘getFile’:
          getFile cfg@(RemoteConfig {..}) attemptNr remoteFile method
            = go method
            where
                go ::
                  Throws SomeRemoteError =>
                  DownloadMethod fs typ
                  -> Verify (Some (HasFormat fs), RemoteTemp typ)
                go (NeverUpdated {..})
                  = do { cfgLogger $ LogDownloading remoteFile;
                         .... }
                go (CannotUpdate {..})
                  = do { cfgLogger $ LogCannotUpdate remoteFile cannotUpdateReason;
                         .... }
                go (Update {..})
                  = do { cfgLogger $ LogUpdating remoteFile;
                         .... }
                headers :: [HttpRequestHeader]
                headers = httpRequestHeaders cfg attemptNr
                ....

src/Hackage/Security/Client/Repository/Remote.hs:382:5: warning: [-Wredundant-constraints]
    • Redundant constraint: Throws SomeRemoteError
    • In the type signature for:
           download :: Throws SomeRemoteError =>
                       HasFormat fs f -> Verify (Some (HasFormat fs), RemoteTemp typ)
      In an equation for ‘getFile’:
          getFile cfg@(RemoteConfig {..}) attemptNr remoteFile method
            = go method
            where
                go ::
                  Throws SomeRemoteError =>
                  DownloadMethod fs typ
                  -> Verify (Some (HasFormat fs), RemoteTemp typ)
                go (NeverUpdated {..})
                  = do { cfgLogger $ LogDownloading remoteFile;
                         .... }
                go (CannotUpdate {..})
                  = do { cfgLogger $ LogCannotUpdate remoteFile cannotUpdateReason;
                         .... }
                go (Update {..})
                  = do { cfgLogger $ LogUpdating remoteFile;
                         .... }
                headers :: [HttpRequestHeader]
                headers = httpRequestHeaders cfg attemptNr
                ....
Preprocessing test suite 'TestSuite' for hackage-security-0.5.1.0...
[1 of 6] Compiling TestSuite.Util.StrictMVar ( tests/TestSuite/Util/StrictMVar.hs, dist/build/TestSuite/TestSuite-tmp/TestSuite/Util/StrictMVar.o )
[2 of 6] Compiling TestSuite.PrivateKeys ( tests/TestSuite/PrivateKeys.hs, dist/build/TestSuite/TestSuite-tmp/TestSuite/PrivateKeys.o )
[3 of 6] Compiling TestSuite.InMemRepo ( tests/TestSuite/InMemRepo.hs, dist/build/TestSuite/TestSuite-tmp/TestSuite/InMemRepo.o )
[4 of 6] Compiling TestSuite.InMemCache ( tests/TestSuite/InMemCache.hs, dist/build/TestSuite/TestSuite-tmp/TestSuite/InMemCache.o )
[5 of 6] Compiling TestSuite.InMemRepository ( tests/TestSuite/InMemRepository.hs, dist/build/TestSuite/TestSuite-tmp/TestSuite/InMemRepository.o )

tests/TestSuite/InMemRepository.hs:48:1: warning: [-Wredundant-constraints]
    • Redundant constraint: Throws SomeRemoteError
    • In the type signature for:
           getRemote :: Throws SomeRemoteError =>
                        InMemRepo
                        -> InMemCache
                        -> AttemptNr
                        -> RemoteFile fs typ
                        -> Verify (Some (HasFormat fs), InMemFile typ)
[6 of 6] Compiling Main             ( tests/TestSuite.hs, dist/build/TestSuite/TestSuite-tmp/Main.o )

tests/TestSuite.hs:25:1: error:
    Failed to load interface for ‘TestSuite.HttpMem’
    Use -v to see a list of the files searched for.

tests/TestSuite.hs:31:1: error:
    Failed to load interface for ‘TestSuite.JSON’
    Use -v to see a list of the files searched for.
```

**Comment by hvr** [comment]
looks a lot like #157 which is fixed in `master` already

**Comment by hvr** [comment]
I've just created a `cabal sdist` tarball from current `master`, and ran `cabal new-build --enable-test && dist-newstyle/build/hackage-security-0.5.1.0/build/TestSuite/TestSuite` in it, and it passed. So consider this fixed in Git.

**Comment by edsko** [comment]
Closed as a duplicate of #157. Thanks for the report!
